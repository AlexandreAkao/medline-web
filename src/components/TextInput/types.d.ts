interface ITextInputProps {
  label?: string;
  placeholder?: string;
  maxLength?: number;
  isPassword?: boolean;
  mask?: import('utils/mask').Mask | string | (string | RegExp)[];
  onChange?: (event: import('react').ChangeEvent<HTMLInputElement>) => void;
  Icon?: import('react-icons').IconType;
  value?: string | number;
}
