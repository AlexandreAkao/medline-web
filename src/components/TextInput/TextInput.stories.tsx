import { ComponentStory, ComponentMeta } from '@storybook/react';
import { FaUserAlt, FaCalendarAlt, FaLock } from 'react-icons/fa';

import TextInput from 'components/TextInput';
import colors from 'styles/colors';

export default {
  title: 'TextInput',
  args: {
    placeholder: 'Placeholder',
  },
  decorators: [
    Story => (
      <div
        style={{
          background: colors.primary,
          width: '100%',
          height: '100%',
          padding: '50px',
        }}
      >
        <Story />
      </div>
    ),
  ],
  component: TextInput,
} as ComponentMeta<typeof TextInput>;

const Template: ComponentStory<typeof TextInput> = args => <TextInput {...args} />;

export const WithIcon = Template.bind({});
WithIcon.args = {
  Icon: FaUserAlt,
  onChange: event => console.log(event),
};

export const WithoutIcon = Template.bind({});
WithoutIcon.args = {};

export const WithLabel = Template.bind({});
WithLabel.args = {
  label: 'Nome',
  onChange: event => console.log(event),
};

export const MaxLength = Template.bind({});
MaxLength.args = {
  maxLength: 20,
};

export const Password = Template.bind({});
Password.args = {
  Icon: FaLock,
  placeholder: 'Senha',
  isPassword: true,
};

export const Date = Template.bind({});
Date.args = {
  mask: '99/99/9999',
  Icon: FaCalendarAlt,
  placeholder: 'Data',
  maxLength: 20,
};

export const Cpf = Template.bind({});
Cpf.args = {
  mask: '999.999.999-99',
  placeholder: 'CPF',
  Icon: FaUserAlt,
};
