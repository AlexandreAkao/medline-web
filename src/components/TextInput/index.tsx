import { useState, useMemo } from 'react';
import InputMask, { Props } from 'react-input-mask';
import { FaEye, FaEyeSlash } from 'react-icons/fa';

import {
  Container,
  LabelTitle,
  TextInputContainer,
  TextInputStyled,
} from 'components/TextInput/styles';
import { Mask } from 'utils/mask';

function TextInput({
  label,
  placeholder,
  maxLength,
  isPassword,
  mask = Mask.Empyt,
  onChange,
  Icon,
  value,
  ...rest
}: ITextInputProps) {
  const [showPassword, setShowPassword] = useState(false);
  const typeInput = useMemo(
    () => (isPassword && !showPassword ? 'password' : 'text'),
    [isPassword, showPassword],
  );

  const toggleShowPassword = () => {
    setShowPassword(prev => !prev);
  };

  const EyeIcon = useMemo(() => (showPassword ? FaEyeSlash : FaEye), [showPassword]);

  return (
    <Container>
      {label && <LabelTitle data-testid="test-input-label">{label}</LabelTitle>}
      <TextInputContainer>
        {Icon && <Icon role="img" size={30} />}
        <InputMask
          {...rest}
          mask={mask}
          onChange={onChange}
          placeholder={placeholder}
          type={typeInput}
          value={value}
          role="textbox"
        >
          {(inputProps: Props) => {
            return (
              <TextInputStyled
                {...inputProps}
                value={value}
                onChange={onChange}
                maxLength={maxLength}
                data-testid="test-input"
              />
            );
          }}
        </InputMask>
        {isPassword && (
          <EyeIcon data-testid="test-eye-icon" role="img" size={26} onClick={toggleShowPassword} />
        )}
      </TextInputContainer>
    </Container>
  );
}

export default TextInput;
