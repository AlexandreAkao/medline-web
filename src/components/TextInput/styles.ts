import styled from 'styled-components';

import colors from 'styles/colors';
import { getColorAlpha } from 'utils/colors';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
`;

export const LabelTitle = styled.span`
  color: ${colors.white.normal};
  font-weight: 700;
  margin-bottom: 12px;
  font-size: 1rem;
`;

export const TextInputContainer = styled.div`
  border: 0;
  width: 100%;
  height: 44px;
  border-radius: 8px;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: ${colors.white.normal};

  svg {
    margin: 10px;
  }

  svg:first-child {
    border-right: 1px solid ${getColorAlpha(colors.grey.normal, 0.5)};
    padding-right: 10px;
    color: ${colors.primary};
  }
`;

export const TextInputStyled = styled.input`
  width: 100%;
  height: 100%;
  border: 0;
  font-size: 1.1rem;
  padding: 0 16px;
  color: ${colors.black.soft};
  font-weight: 500;
  border-radius: 8px;

  ::placeholder {
    color: ${colors.grey.normal};
  }

  & + svg {
    color: ${colors.primary};
    cursor: pointer;
    padding: 2px;
  }
`;
