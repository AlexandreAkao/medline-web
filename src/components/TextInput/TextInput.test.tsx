import { render, screen } from '@testing-library/react';
import { composeStories } from '@storybook/testing-react';
import userEvent from '@testing-library/user-event';
import { vi } from 'vitest';

import * as stories from 'components/TextInput/TextInput.stories';
import TextInput from 'components/TextInput';

const { WithIcon, WithoutIcon, WithLabel, MaxLength, Password, Date, Cpf } =
  composeStories(stories);

describe('TextInput', () => {
  const renderComponent = (customProps: Partial<ITextInputProps> = {}) => {
    const props: ITextInputProps = {
      placeholder: 'any_placeholder',
      ...customProps,
    };
    return render(<TextInput {...props} />);
  };

  it('should match with snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render a text input with icon', () => {
    render(<WithIcon />);
    const iconElement = screen.getByRole('img');
    const labelElement = screen.queryByTestId('test-input-label');

    expect(iconElement).toBeInTheDocument();
    expect(labelElement).not.toBeInTheDocument();
  });

  it('should render a text input without icon', () => {
    render(<WithoutIcon />);
    const iconElement = screen.queryByRole('img');
    const labelElement = screen.queryByTestId('test-input-label');

    expect(iconElement).not.toBeInTheDocument();
    expect(labelElement).not.toBeInTheDocument();
  });

  it('should render a text input with label', () => {
    render(<WithLabel />);
    const labelElement = screen.getByTestId('test-input-label');
    const labelText = screen.getByText('Nome');

    expect(labelElement).toBeInTheDocument();
    expect(labelText).toBeInTheDocument();
  });

  it('should render a text input with max length', async () => {
    render(<MaxLength />);
    const labelElement = screen.queryByTestId('test-input-label');
    const labelText = screen.queryByText('Nome');
    const inputElement = screen.getByTestId('test-input');

    await userEvent.type(inputElement, 'any text with more than 20 characters');

    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toHaveValue('any text with more t');
    expect(labelElement).not.toBeInTheDocument();
    expect(labelText).not.toBeInTheDocument();
  });

  it('should render a text input as password', async () => {
    render(<Password />);
    const inputElement = screen.getByTestId('test-input');
    const eyeIconElement = screen.getByTestId('test-eye-icon');

    await userEvent.type(inputElement, 'any password');

    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toHaveValue('any password');
    expect(inputElement).toHaveAttribute('type', 'password');

    await userEvent.click(eyeIconElement);

    expect(inputElement).toHaveAttribute('type', 'text');
  });

  it('should render a text input as date', async () => {
    render(<Date />);
    const inputElement = screen.getByTestId('test-input');

    await userEvent.click(inputElement);

    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toHaveDisplayValue('__/__/____');

    await userEvent.type(inputElement, 'aaaaaaa', {
      skipClick: true,
    });

    expect(inputElement).toHaveValue('__/__/____');

    await userEvent.type(inputElement, '999999999999999999', {
      skipClick: true,
    });

    expect(inputElement).toHaveValue('99/99/9999');
  });

  it('should render a text input as CPF', async () => {
    render(<Cpf />);
    const inputElement = screen.getByTestId('test-input');

    await userEvent.click(inputElement);

    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toHaveDisplayValue('___.___.___-__');

    await userEvent.type(inputElement, 'aaaaaaa', {
      skipClick: true,
    });

    expect(inputElement).toHaveValue('___.___.___-__');

    await userEvent.type(inputElement, '999999999999999999', {
      skipClick: true,
    });

    expect(inputElement).toHaveValue('999.999.999-99');
  });

  it('should trigger function on change', async () => {
    const onChange = vi.fn();
    renderComponent({ onChange });
    const inputElement = screen.getByTestId('test-input');

    await userEvent.type(inputElement, 'any text');

    expect(inputElement).toHaveValue('any text');
    expect(onChange).toBeCalled();
  });
});
