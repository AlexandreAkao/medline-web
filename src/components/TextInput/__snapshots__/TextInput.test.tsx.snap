// Vitest Snapshot v1

exports[`TextInput > should match with snapshot 1`] = `
<div>
  <div
    class="sc-bczRLJ itlDIl"
  >
    <div
      class="sc-dkzDqf cKodea"
    >
      <input
        class="sc-hKMtZM fIniom"
        data-testid="test-input"
        placeholder="any_placeholder"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
